sdk/cosmos/azure_data_cosmos/src/routing/location_cache.rs

```diff
@@ -163,35 +275,96 @@ impl LocationCache {
         }
     }
 
-    pub fn resolve_service_endpoint(
-        &self,
-        location_index: usize,
-        operation: RequestOperation,
-    ) -> String {
+    /// Resolves the appropriate service endpoint for a request.
+    ///
+    /// # Summary
+    /// Determines which endpoint should handle the request based on operation type (read vs write),
+    /// location index routing preference, multi-write support, and preferred location settings.
+    /// For requests not using preferred locations or writes without multi-write support, routes to
+    /// write locations directly. Otherwise, selects from ordered read/write endpoint lists based
+    /// on location index (with wraparound via modulo). Returns default endpoint if no regions
+    /// are configured.
+    ///
+    /// # Arguments
+    /// * `request` - The Cosmos DB request requiring endpoint resolution
+    ///
+    /// # Returns
+    /// The resolved endpoint URL as a String
+    pub fn resolve_service_endpoint(&self, request: &CosmosRequest) -> String {
         // Returns service endpoint based on index, if index out of bounds or operation not supported, returns default endpoint
-        if operation == RequestOperation::Write && !self.locations_info.write_endpoints.is_empty() {
-            self.locations_info
-                .write_endpoints
-                .get(location_index)
-                .cloned()
-                .unwrap_or_else(|| self.default_endpoint.clone())
-        } else if operation == RequestOperation::Read
-            && !self.locations_info.read_endpoints.is_empty()
+        let location_index = request.request_context.location_index_to_route.unwrap_or(0) as usize;
+        let mut location_endpoint_to_route = self.default_endpoint.clone();
```

Comment:
nit: This is always going to clone the `default_endpoint` value, even if you end up finding a regional endpoint.

I see two good solutions:

1. Change all the various places you set `location_endpoint_to_route` to just return the endpoint right away, thus removing the need for the variable altogether
2. Change `location_endpoint_to_route` to be an `Option<...>`, and then at the end return `location_endpoint_to_route.unwrap_or_else(|| self.default_endpoint.clone())`